syntax = "proto3";
package qmtl.dagmanager;

message DiffRequest {
  string strategy_id = 1;
  string dag_json = 2;
}

message DiffChunk {
  map<string,string> queue_map = 1;
  string sentinel_id = 2;
}

service DiffService {
  rpc Diff(DiffRequest) returns (stream DiffChunk);
}

message TagQueryRequest {
  repeated string tags = 1;
  int64 interval = 2;
}

message TagQueryReply {
  repeated string queues = 1;
}

service TagQuery {
  rpc GetQueues(TagQueryRequest) returns (TagQueryReply);
}

message CleanupRequest {
  string strategy_id = 1;
}

message CleanupResponse {}

message QueueStatsRequest {
  string filter = 1;
}

message QueueStats {
  map<string,int64> sizes = 1;
}

service AdminService {
  rpc Cleanup(CleanupRequest) returns (CleanupResponse);
  rpc GetQueueStats(QueueStatsRequest) returns (QueueStats);
  rpc RedoDiff(RedoDiffRequest) returns (DiffResult);
}

message RedoDiffRequest {
  string sentinel_id = 1;
  string dag_json = 2;
}

message DiffResult {
  map<string,string> queue_map = 1;
  string sentinel_id = 2;
  rpc RedoDiff(RedoDiffRequest) returns (DiffResult);
}

message RedoDiffRequest {
  string sentinel_id = 1;
  string dag_json = 2;
}

message DiffResult {
  map<string,string> queue_map = 1;
  string sentinel_id = 2;
}

message PingRequest {}
message PingReply {}

service HealthCheck {
  rpc Ping(PingRequest) returns (PingReply);
}
